### PR TITLE
feat(worker): restart event thread to avoid memory leak

### PR DESCRIPTION
**Pull request by chronolaw**
KAG-5319

Related doc: https://docs.google.com/document/d/14Xu8HtHZNF-of49LrLiyETlG_uE6tYa4xdqIPhoD3m0/edit

### CODE REVIEW

File: lualib/resty/events/worker.lua [hunk 1]
```diff
@@ -32,6 +32,7 @@ local decode = codec.decode
 local cjson_encode = cjson.encode
 
 local EVENTS_COUNT_LIMIT = 100
+local EVENTS_POP_LIMIT = 5000
```
Comment (outsinre) [inline lualib/resty/events/worker.lua:35]: timerng currently uses a fixed value `5000`. I wonder if this value is too high or too low.

**Review by outsinre — DISMISSED** [review]

File: lualib/resty/events/worker.lua (hunk 1 above)
Comment (chronolaw) [reply lualib/resty/events/worker.lua:35]: For now let's keep consistent with it.

Comment (outsinre) [reply lualib/resty/events/worker.lua:35]: Before merge it, I would suggest send this patch to user for test, along with debug.traceback.

**Review by outsinre — APPROVED** [review]

**Comment by chobits** [comment]
Ok cool, i wanna create a scenario to leak memory in kong, giving me some time

**Review by bungle — CHANGES_REQUESTED** [review]

File: lualib/resty/events/worker.lua [hunk 2]
```diff
@@ -236,6 +239,13 @@ local function events_thread(self)
...
 
+        counter = counter + 1
+
+        -- exit and restart timer to avoid memory leak
+        if counter >= EVENTS_POP_LIMIT then
+            break
+        end
+
         -- yield, not block other threads
         sleep(0)
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply lualib/resty/events/worker.lua:250]: I think that it is not necessary, since after `break` nginx will reschedule a new timer, it will yield.

**Comment by chobits** [comment]
I gave a Poc to trigger this leak in kong, https://github.com/Kong/kong-ee/pull/10190

File: lualib/resty/events/worker.lua (hunk 1 above)
Comment (outsinre) [reply lualib/resty/events/worker.lua:35]: @chronolaw just checked with @ADD-SP , the default value on timerng side is `2000`. There is a misinfo on the readme.

https://github.com/Kong/lua-resty-timer-ng/blob/aec8b5f109e30cabf2f453222a374dc0bca56157/lib/resty/timerng/constants.lua#L17

Comment (chronolaw) [reply lualib/resty/events/worker.lua]: done.

**Review by outsinre — APPROVED** [review]

File: lualib/resty/events/worker.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (chronolaw) [reply lualib/resty/events/worker.lua:250]: @bungle , could you take a look again? thanks.

**Review by chobits — APPROVED** [review]

**Review by ADD-SP — APPROVED** [review]

**Review by bungle — APPROVED** [review]

**Pull request merged**